src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+        readQuotedField(value, buf);
+        assertChar(',', buf);
+        skipWhitespaceIfAny(buf);
+
+        if (key == "'descr'")
+            descr = value;
+        else if (key == "'fortran_order'")
+        {
+            if (value != "false")
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Fortran order is not supported");
+        }
+        else if (key == "'shape'")
+            shape = value;
+    }
+
+    if (shape.empty() || descr.empty())
+        throw Exception(ErrorCodes::INCORRECT_DATA, "npy file header doesn't contain required field 'shape' or 'descr'");
+
+    size_t read_bytes = buf.count() - header_start;
+    if (read_bytes > header_length)
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Header size is incorrect");
+
+    /// Ignore remaining header data.
+    buf.ignore(header_length - read_bytes);
+
+    if (descr[0] == '\'')
+        descr = descr.substr(1, descr.length() - 1);
+    if (descr[descr.length() - 1] == '\'')
+        descr = descr.substr(0, descr.length() - 1);
+
+    if (shape[0] == '\'')
+        shape = shape.substr(1, shape.length() - 1);
+    if (shape[shape.length() - 1] == '\'')
+        shape = shape.substr(0, shape.length() - 1);
+
+    std::unordered_map<String, String> header_data;
+    header_data["shape"] = shape;
+    header_data["descr"] = descr;
+
+    return header_data;
+}
+
+int parseStringSize(const std::string type)
+{
+    int size;
+    try
+    {
+        size = std::stoi(type.substr(2, type.length() - 2));
+        return size;
+    }
+    catch (...)
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid data type");
+    }
+}
+
+NpyRowInputFormat::NpyRowInputFormat(ReadBuffer & in_, Block header_, Params params_)
+    : IRowInputFormat(std::move(header_), in_, std::move(params_))
+{
+    header = parseHeader(*in);
+    endian = endianOrientation(header["descr"]);
+    shape = parseShape(header["shape"]);
+    nestedType = parseType(header["descr"]);
+
+    if (isString(nestedType))
+        sizeForStrings = parseStringSize(header["descr"]);
+}
+
+void NpyRowInputFormat::readRows(MutableColumns & columns)
+{
+    auto & column = columns[0];
+    IColumn * current_column = column.get();
+    size_t elements_in_current_column = 1;
+    for (size_t i = 1; i != shape.size(); ++i)
+    {
+        auto & array_column = assert_cast<ColumnArray &>(*current_column);
+        /// Fill offsets of array columns.
+        for (size_t j = 0; j != elements_in_current_column; ++j)
+            array_column.getOffsets().push_back(array_column.getOffsets().back() + shape[i]);
+        current_column = &array_column.getData();
+        elements_in_current_column *= shape[i];
+    }
+
+    for (size_t i = 0; i != elements_in_current_column; ++i)
+        readValueAndinsertIntoColumn(current_column->getPtr());
+}
+
+void NpyRowInputFormat::readValueAndinsertIntoColumn(MutableColumnPtr column)
+{
+    if (auto * column_int8 = typeid_cast<ColumnInt8 *>(column.get()))
```

Comment:
It will be better to use `switch` over `nested_type->getTypeId()` (`nested_type` from previous comment).
And ClickHouse data type can be different from Numpy data type, and we want to support it. For example, we can have ClickHouse type `UInt32` and Numpy type `u16` (uint16_t), and we want to be able to read uint16_t values into UInt32 column. To do it, we first need to switch over ClickHouse data type, and for each type call appropriate function to parse and insert data. In this function we can iterate over Numpy data types to read proper value and then insert this value (with possible cast) to our column.
As an example you can look at BSONEachRowRowInputFormat class, it has the same logic (see `readField`, `readAndInsertInteger`, `readAndInsertString` functions)